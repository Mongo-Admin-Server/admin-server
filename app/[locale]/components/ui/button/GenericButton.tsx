import { useMemo } from 'react';

import styles from './button.module.scss';

import SvgIcon from '@components/ui/icon/SvgIcon';

interface ButtonProps {
  children?: React.ReactNode;
  onClick?: () => void;
  type?: 'button' | 'submit' | 'reset';
  size?: 'small' | 'auto' | 'full';
  disabled?: boolean;
  padding?: string;
  outline?: boolean;
  center?: boolean;
  transparent?: boolean;
  icon_name?: string;
  variant?: 'danger' | 'warning' | 'success' | 'info'; 
}

const GenericButton = ({
  children,
  onClick,
  type = 'button',
  disabled = false,
  size = 'full',
  padding,
  outline = false,
  center,
  transparent,
  variant,
  icon_name,
}: ButtonProps) => {

  const buttonClass = useMemo(() => {
    const dimensions = {
      small: styles["small"],
      auto: styles["auto"],
      full: styles["full"],
    };
    const variants = {
      danger: styles["danger"],
      warning: styles["warning"],
      success: styles["success"],
      info: styles["info"],
    };

    const _arr = [styles["button"]];
    _arr.push(dimensions[size]);
    if (outline) _arr.push(styles.outline);
    if (center) _arr.push(styles.center);
    if (transparent) _arr.push(styles.transparent);
    if (variant) _arr.push(variants[variant]);

    return _arr.join(' ');
  }, [size, outline, center, transparent, variant]);

  const icon = useMemo(() => {
    if (icon_name) {
      return <SvgIcon icon_name={icon_name} className={styles.icon} />;
    }
  }, [icon_name]);

  return (
    <button
      className={buttonClass}
      onClick={onClick}
      type={type}
      disabled={disabled}
      style={{ padding }}
    >
      {icon}
      {children}
    </button>
  );
};

export default GenericButton;
