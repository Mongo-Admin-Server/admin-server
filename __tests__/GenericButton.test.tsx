/* eslint-disable no-undef */
import { fireEvent, render, screen } from '@testing-library/react'
import GenericButton from '@/app/[locale]/components/ui/button/GenericButton'

const mock = {
  label: 'Generic button'
}
describe("GenericButton", () => {
  it('renders', () => {
    const { container } = render(<GenericButton >{mock.label}</GenericButton>)
    expect(container).toMatchSnapshot();
  })

  it('has working props', () => {
    const handleClick = jest.fn();
    render(
      <GenericButton 
        type='button' 
        onClick={handleClick} 
        disabled={false}
      >
        {mock.label}
      </GenericButton>
    );
    const button = screen.getByRole('button');
    expect(button).toBeInTheDocument();
    expect(button).toHaveTextContent('Generic button')
    expect(button).toHaveClass('button')
    fireEvent.click(button)
    expect(handleClick).toHaveBeenCalledTimes(1);
  })

  it('can be disabled', () => {
    const handleClick = jest.fn();
    render(
      <GenericButton 
        disabled={true} 
        onClick={handleClick}
      >
        {mock.label}
      </GenericButton>
    )
    const button = screen.getByRole('button')
    expect(button).toHaveAttribute('disabled')
    fireEvent.click(button)
    expect(handleClick).toHaveBeenCalledTimes(0)
  })
})