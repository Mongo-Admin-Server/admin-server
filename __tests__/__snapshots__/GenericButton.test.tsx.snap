// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenericButton renders 1`] = `
<div>
  <button
    class="button full"
    type="button"
  >
    Generic button
  </button>
</div>
`;
